src/components/pickers/AutoFill/BaseAutoFill.tsx

```diff
@@ -0,0 +1,157 @@
+
+  constructor(props: IBaseAutoFillProps) {
+    super(props);
+    this.state = {
+      value: ''
+    };
+  }
+
+  get cursorLocation(): number {
+    let inputElement = this._inputElement;
+    if (inputElement.selectionDirection !== 'forward') {
+      return inputElement.selectionEnd;
+    } else {
+      return inputElement.selectionStart;
+    }
+  }
+
+  get isValueSelected() {
+    return this.inputElement.selectionStart !== this.inputElement.selectionEnd;
+  }
+
+  get value(): string {
+    return this.state.value;
+  }
+
+  get selectionStart(): number {
+    return this._inputElement.selectionStart;
+  }
+
+  get selectionEnd(): number {
+    return this._inputElement.selectionEnd;
+  }
+
+  get inputElement(): HTMLInputElement {
+    return this._inputElement;
+  }
+
+  public componentDidUpdate() {
+    let { value } = this.state;
+    let { suggestedDisplayValue } = this.props;
+    let differenceIndex = 0;
+    if (value && suggestedDisplayValue && this._doesTextStartWith(suggestedDisplayValue, value)) {
+      while (differenceIndex < value.length && value[differenceIndex].toLocaleLowerCase() === suggestedDisplayValue[differenceIndex].toLocaleLowerCase()) {
+        differenceIndex++;
+      }
+      if (differenceIndex > 0) {
+        this._inputElement.setSelectionRange(differenceIndex, suggestedDisplayValue.length, 'backward');
+      }
+    }
+  }
+
+  public render() {
+    let { value } = this.state;
+    let {
+      ariaDescription,
+      ariaActiveDescendant,
+      ariaOwns,
+      ariaExpanded,
+      ariaHasPopup,
+      suggestedDisplayValue
+    } = this.props;
+
+    let displayValue = this.state.value;
+    if (this._doesTextStartWith(suggestedDisplayValue, value)) {
+      displayValue = suggestedDisplayValue;
+    }
+    const nativeProps = getNativeProps(this.props, inputProperties);
+    return React.createElement(tag,
+      assign(
+        {},
+        nativeProps,
+        {
+          'ref': this._resolveRef('_inputElement'),
+          'value': displayValue,
+          'autoCapitalize': 'off',
+          'autoComplete': 'off',
+          'onChange': this._onChange,
+          'onKeyDown': this._onKeyDown,
+          'aria-activedescendant': ariaActiveDescendant,
+          'aria-owns': ariaOwns,
+          'aria-expanded': ariaExpanded,
+          'aria-haspopup': ariaHasPopup,
+          'aria-describedby': ariaDescription
+        }));
+  }
+
+  public focus() {
+    this._inputElement.focus();
+  }
+
+  public clear() {
+    this._updateValue('');
+  }
+
+  @autobind
+  private _onKeyDown(ev: React.KeyboardEvent<HTMLElement>) {
+    switch (ev.which) {
+      case KeyCodes.backspace:
+        this._onBackspace(ev);
+        break;
+    }
+  }
+
+  private _onBackspace(ev: React.KeyboardEvent<HTMLElement>) {
+    let { value } = this.state;
+    if (ev.target === this._inputElement) {
+      if (value && value.length > 0) {
+        this._updateValue(value.substring(0, value.length - 1));
```

Comment:
So they are different, but which one to use isn't really clear. Substr is substr(start, length) and substring is substring(startindex, endindex). Substr supports a negative start and will go from the end of the string. So with substr you can do ('hello there').substr(-4, 2) will produce "he".